spec.bs

```diff
@@ -0,0 +1,520 @@
+
+  <h2 id="collect-page-topics-calculation-input-data-header">Collect page topics calculation input data</h2>
+  <div algorithm>
+    To <dfn>collect page topics calculation input data</dfn>, given a {{Document}} |document|:
+
+    1. If |document|'s [=node navigable=] is a [=prerendering navigable=], then append the following steps to |document|'s [=post-prerendering activation steps list=] and return. Else, run the following steps [=in parallel=]:
+      1. Let |documentId| be |document|'s [=document-id-header/document ID=].
+      1. If user agent's [=user agent/topics history storage=] contains a [=topics history entry=] whose [=topics history entry/document id=] is |documentId|, return.
+      1. Let |topicsHistoryEntry| be a [=topics history entry=].
+      1. Set |topicsHistoryEntry|'s [=topics history entry/document id=] to |documentId|.
+      1. Set |topicsHistoryEntry|'s [=topics history entry/topics calculation input data=] to the [=determine-topics-calculation-input-data-header/topics calculation input data=] for |document|.
+      1. Set |topicsHistoryEntry|'s [=topics history entry/time=] to |document|'s [=relevant settings object=]'s [=current relative timestamp=].
+      1. [=list/Append=] |topicsHistoryEntry| to user agent's [=user agent/topics history storage=].
+  </div>
+
+  <h2 id="collect-topics-caller-origin-header">Collect topics caller origin</h2>
+  <div algorithm>
+    To <dfn>collect topics caller origin</dfn>, given a {{Document}} |document| and an [=origin=] |callerOrigin|:
+
+    1. Run the following steps [=in parallel=]:
+      1. Let |documentId| be |document|'s [=document-id-header/document ID=].
+      1. If user agent's [=user agent/topics history storage=] does not contain a [=topics history entry=] whose [=topics history entry/document id=] is |documentId|, return.
+      1. Let |topicsHistoryEntry| be the [=topics history entry=] in user agent's [=user agent/topics history storage=] whose [=topics history entry/document id=] is |documentId|.
+      1. [=set/Append=] |callerOrigin| to |topicsHistoryEntry|'s [=topics caller origins=].
+  </div>
+
+  <h2 id="periodically-calculate-user-topics-header">Periodically calculate user topics</h2>
+  <div algorithm>
+    To <dfn>calculate user topics</dfn>, perform the following steps every 7 days:
+    1. Let |now| be the [=monotonic clock=]'s [=monotonic clock/unsafe current time=].
+    1. If either user agent's [=user agent/model=] or [=user agent/taxonomy=] isn't available:
+        1. Let |epoch| be an [=epoch=] struct with default initial field values.
+        1. Set |epoch|'s [=epoch/time=] to |now|.
+        1. [=list/Append=] |epoch| to user agent's [=user agent/user topics state=]'s [=user topics state/epochs=].
+        1. If user agent's [=user agent/user topics state=]'s [=user topics state/epochs=] has more than 4 entries, remove the oldest epoch (i.e. the epoch with index 0), then return.
```

Comment:
Is "then return" in the right place? This currently says that if the UA has 4 or fewer epochs, then it should continue on to the steps below, some of which use the taxonomy that step 2 might have concluded it didn't have.

Reply:
You're right. Thanks for catching! Fixed.